### ISSUE TITLE
Softmax Node has useless input operand (selected)?

### ISSUE THREAD

**Issue by et-nivard** [pytorch/glow#5919]
Softmax Node has a second input named "selected" https://github.com/pytorch/glow/blob/master/tools/ClassGen/NodeGen.cpp#L374 which doesn't seem to be used at all. Actually, Softmax instruction doesn't have this extra input https://github.com/pytorch/glow/blob/master/tools/ClassGen/InstrGen.cpp#L300 and there is no specific lowering between them. Only in the case of gradient version it seems to have a usage.
Could we remove this apparently useless input and cleanup the corresponding OperatorTest? Or maybe I'm missing something?

**Comment by jfix71** [comment]
I believe this is used when doing training. E.g. see IRGen for SoftMaxGrad

https://github.com/pytorch/glow/blob/d54671836d6a00eb86d5c117707a557d1d4945de/lib/IR/IRGen.cpp#L222-L239

**Comment by et-nivard** [comment]
Yes, I know. But why does SoftMax Node also contain this input operand? Is it required due to some relationship or flow between SoftMax and SoftMaxGrad? If not, why not removing it? It seems like redundant code to me.

**Comment by jfix71** [comment]
Yes, after differentiation it is also passed as an input into `SoftMaxGradNode`, and is used there. See screenshot below of an example dot, from MLTest's `classifyPlayerSport`:

<img width="674" alt="smg" src="https://user-images.githubusercontent.com/1198212/156858984-3a5356aa-82d6-47bc-8c8b-846997fd0042.png">

We include it in the `SoftMaxNode` so that when we autogenerate the `SoftMaxGradNode` it's there for use by the SoftMaxGrad kernel.
https://github.com/pytorch/glow/blob/4c1b7875e32266917f608667ece99eb27aa4e9b0/lib/Backends/Interpreter/InterpreterNodes.cpp#L2253-L2269

**Comment by et-nivard** [comment]
OK, got it. Thanks.
